content/movie-review-pt2.md

```diff
@@ -0,0 +1,522 @@
+
+For our `MovieAccountState` we are storing four fields. We will allocate 1 byte each for the `is_initialized` and `rating`. For both `title` and `description` we will allocate space equal to 4 bytes plus the length of the string. The additional 4 bytes is prefix that stores the actual length of the string input by the user. This prefix is used to find where the next field is located in the array of bytes (i.e. how many bytes after after `title` to find `description`).
+
+We’ll store the space we calculate in a variable called `account_len`. We then calculate the rent required for the space we need using the `minimum_balance` function from the `rent` module of the `solana_program` crate.
+
+```rust
+	// Calculate account size required for struct MovieAccountState
+  let account_len: usize = 1 + 1 + (4 + title.len()) + (4 + description.len());
+
+	// Calculate rent required
+  let rent = Rent::get()?;
+  let rent_lamports = rent.minimum_balance(account_len);
+```
+
+## Program Derived Addresses (PDA)
+
+To create an account, we also need to have an address to assign the account. For program owned accounts, this will be a program derived address (PDA) found using the `find_program_address` function. As the name implies, PDAs are derived using the program ID (address of the program creating the account) and an optional list of “seeds”.  Optional seeds are additional inputs used in the function to derive the PDA.  Providing additional seeds to derive a PDA allows us to create an arbitrary number of accounts the program may own.
+
+For our movie review program, we will two additional seeds to derive the PDA for each new account. We will use the initializer’s publickey and the movie title as the optional seeds. Setting up the PDA this way restrict each user to only one review for any one movie title. However, it still allows the same user to review with movies with different titles and different users to review the movie with the same title.
+
+There is one more seed that is needed to derive a PDA. This seed is referred to as the “bump seed” and is a number between 255-0. It is a requirement that PDAs do not have a corresponding private key. This is done by rederiving the PDA using a new “bump seed” until a valid PDA is found. The `find_program_addres` function tries to find a PDA using the optional seeds provided, the program ID, and the “bump seed” starting from 255. If the output is not a valid PDA, then the function decreases the bump by 1 and tries again (255, 254, 253, etc). Once a valid PDA is found, the function returns both the PDA and the bump that was used to derive the PDA.
```

Comment:
@ZYJLiu rather than stating:

"This is done by rederiving the PDA using ..."

is it acceptable to instead say:

"This is done by deriving the PDA again using ..."

Reply:
updated